main.go

```diff
@@ -3,45 +3,147 @@ package main
+type Clock struct {
+	instant time.Time
+}
 
-	ReadFile(ctx context.Context, readBucket string, readLocation string) ([]byte, error)
+func (this *Clock) Now() time.Time {
+	if this == nil {
+		return time.Now()
+	}
+	return this.instant
 }
 
 type Firehose struct {
-	processSettings ProcessSettings
-	storageBackend  StorageBackend
-	hedwigConsumer  *hedwig.QueueConsumer
+	processSettings       ProcessSettings
+	storageBackendCreator StorageBackendCreator
+	hedwigConsumer        *hedwig.QueueConsumer
+	hedwigFirehose        *hedwig.Firehose
+	flushLock             sync.Mutex
+	flushCh               chan error
+	messageCh             chan ReceivedMessage
+	listenRequest         hedwig.ListenRequest
+	clock                 *Clock
+}
+
+func (fp *Firehose) flushCron(ctx context.Context) {
+	errChannelMapping := make(map[hedwig.MessageTypeMajorVersion][]chan error)
+	writerMapping := make(map[hedwig.MessageTypeMajorVersion]io.WriteCloser)
+	currentTime := fp.clock.Now()
+	timerCh := time.After(time.Duration(fp.processSettings.FlushAfter) * time.Second)
+	// go through all msgs and write to msgtype folder
+	for {
+		select {
+		case <-timerCh:
+			// close all writers and associated errChannels
+			for key, writer := range writerMapping {
+				err := writer.Close()
+				errChannels := errChannelMapping[key]
+				if err != nil {
+					for _, errCh := range errChannels {
+						errCh <- err
+					}
+				} else {
+					for _, errCh := range errChannels {
+						close(errCh)
+					}
+				}
+			}
+			// start a new flushcron go routine, as this one is done
+			go fp.flushCron(ctx)
+			return
+		case messageAndChan := <-fp.messageCh:
+			message := messageAndChan.message
+			errCh := messageAndChan.errCh
+			key := hedwig.MessageTypeMajorVersion{
+				MessageType:  message.Type,
+				MajorVersion: uint(message.DataSchemaVersion.Major()),
+			}
+			// if writer doesn't exist create in mapping
+			if _, ok := writerMapping[key]; !ok {
+				uploadLocation := fmt.Sprintf("%s/%s/%s/%s", key.MessageType, fmt.Sprint(key.MajorVersion), currentTime.Format("2006/1/2"), fmt.Sprint(currentTime.Unix()))
+				writer, err := fp.storageBackendCreator.CreateWriter(ctx, fp.processSettings.StagingBucket, uploadLocation)
+				if err != nil {
+					errCh <- err
+					continue
+				}
+				writerMapping[key] = writer
+			}
+			msgTypeWriter := writerMapping[key]
+			payload, err := fp.hedwigFirehose.Serialize(message)
+			if err != nil {
+				errCh <- err
+				continue
+			}
+			_, err = msgTypeWriter.Write(payload)
+			if err != nil {
+				errCh <- err
+				continue
+			}
+			errChannelMapping[key] = append(errChannelMapping[key], errCh)
+		}
+	}
+
 }
 
-func (fp *Firehose) WriteMessages() {
+func (fp *Firehose) RunFollower(ctx context.Context) {
+	// run an infinite loop until canceled
+	// and call handleMessage
+	go fp.flushCron(ctx)
+	err := fp.hedwigConsumer.ListenForMessages(ctx, fp.listenRequest)
+	// consumer errored so panic
+	if err != nil {
```

Comment:
would add an exception for `&& err != context.Canceled` here since that's the main way process will shut down.

Reply:
I see so on cancel it exits without panic

Reply:
yep exactly.